### PR TITLE
gmoccapy -add log levels to gmoccapy and getiniinfo

### LINKED ISSUE #2273: Configuration warnings by sim.gmoccapy
Hello, Just to make a good impression on new users (and convey trust for those transitioning to the new release) we should avoid warnings on stdout/stderr. I came across

```
**** GMOCCAPY INFO ****
**** Entering make macro button
found 15 Macros
Macro go_to_position x-pos y-pos z-pos has image link
Image = /home/moeller/linuxcnc/configs/sim.gmoccapy/macros/images/goto_x_y_z.png
get button with image
Macro i_am_lost has image link
Image = /home/moeller/linuxcnc/configs/sim.gmoccapy/macros/images/i_am_lost.png
get button with image
Macro macro_8 has image link
Image = /home/moeller/linuxcnc/configs/sim.gmoccapy/macros/images/macro_8.png
get button with image
**** GMOCCAPY INFO ****
```
Why is this shown? Maybe a debug level wrongly set? The files are existing:
```
-rw-r--r-- 1 moeller moeller  24520 26. Dez 18:38 /home/moeller/linuxcnc/configs/sim.gmoccapy/macros/images/goto_x_y_z.png
-rw-r--r-- 1 moeller moeller  14074 26. Dez 18:38 /home/moeller/linuxcnc/configs/sim.gmoccapy/macros/images/i_am_lost.png
-rw-r--r-- 1 moeller moeller 222151 26. Dez 18:38 /home/moeller/linuxcnc/configs/sim.gmoccapy/macros/images/macro_8.png
```
I could open these files in gimp (macro_8 is very cute). So, maybe just not print this?

```
**** GMOCCAPY INFO ****
**** Place in table
/usr/bin/gmoccapy:1565: DeprecationWarning: Gtk.Table.resize is deprecated
  self.widgets.tbl_DRO.resize(rows, cols)
**** GMOCCAPY INFO ****
```
https://docs.gtk.org/gtk3/method.Table.resize.html mentions that this is deprecated since 3.4 and GtkGrid would resize automatically. Such dynamics may be interesting - after all, reactive web pages are a thing, too. No idea how will this works for us, though.

```
**** GMOCCAPY INFO ****
**** Gremlin Einrichtung gestartet ****
Found an error!
The following information may be useful in troubleshooting:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/gladevcp/hal_gremlin.py", line 164, in do_set_property
    if self.lathe_option:
       ^^^^^^^^^^^^^^^^^
AttributeError: 'HAL_Gremlin' object has no attribute 'lathe_option'
```
The default "sim.gmoccapy" is a 3-axis mill and that is why that lathe_option is not present?!?
Directly underneath it states the following, which may be related (even though it does not look like it):
```
not enough values to unpack (expected 2, got 1)
'btn_sel_next'
'btn_sel_prev'
'btn_select'
'btn_home'
'btn_dir_up'
'btn_jump_to'
not enough values to unpack (expected 2, got 1)
not enough values to unpack (expected 2, got 1)
'IconFileSelection' object has no attribute 'on_btn_jump_to'
not enough values to unpack (expected 2, got 1)
not enough values to unpack (expected 2, got 1)

(gmoccapy:101189): Gtk-CRITICAL **: 22:43:43.955: gtk_tree_view_scroll_to_cell: assertion 'tree_view->priv->tree != NULL' failed

(gmoccapy:101189): Gtk-CRITICAL **: 22:43:44.047: gtk_entry_set_text: assertion 'text != NULL' failed
Attribute error <GParamString 'icon-theme-path'> and <class 'str'> sans 10
Attribute error <GParamString 'icon-theme-name'> and <class 'str'> sans 10
0  =  /usr/bin/gmoccapy
1  =  -ini
2  =  /home/moeller/linuxcnc/configs/sim.gmoccapy/gmoccapy.ini
Entry = trivkins
Entry = coordinates=xyz
found the following coordinates xyz
```
Harmless:
```
**** virtual keyboard program found : <onboard>
14:09:30.842 WARNING Config: mousetweaks GSettings schema not found, mousetweaks integration disabled.
/usr/bin/gmoccapy:3318: DeprecationWarning: Gtk.FontButton.get_font_name is deprecated
  self.notification.set_property('font', self.widgets.fontbutton_popup.get_font_name())
/usr/bin/gmoccapy:381: DeprecationWarning: Gtk.FontButton.set_font_name is deprecated
  self.widgets.fontbutton_popup.set_font_name(self.prefs.getpref("message_font", "sans 10", str))
/usr/bin/gmoccapy:476: DeprecationWarning: Gtk.ColorButton.set_color is deprecated
  self.widgets.abs_colorbutton.set_color(Gdk.color_parse(self.abs_color))
```
which has been deprecated with version 3.4 (according to https://athenajc.gitbooks.io/python-gtk-3-api/content/gtk-group/gtkcolorbutton.html) and we are requesting 3.11+ as a dependency.

Similarly:
```
**** GMOCCAPY INFO ****
**** Entering make touch button
/usr/bin/gmoccapy:972: DeprecationWarning: Gtk.Widget.override_background_color is deprecated
  btn.override_background_color(Gtk.StateFlags.ACTIVE, Gdk.RGBA(1.0, 1.0, 0.0, 1.0))
**** GMOCCAPY INFO ****
```
Maybe just give default settings?
```
**** GMOCCAPY GETINIINFO **** 
[KINS] KINESTYPE is trivkins
**** GMOCCAPY GETINIINFO **** 
No MIN_ANGULAR_VELOCITY entry found in [DISPLAY] of INI file
Using default value of 0.1 degree / min
**** GMOCCAPY GETINIINFO **** 
No DEFAULT_ANGULAR_VELOCITY entry found in [DISPLAY] of INI file
Using default value of 360 degree / min
**** GMOCCAPY GETINIINFO **** 
No MAX_RAPID_OVERRIDE entry found in [DISPLAY] of INI file 
 Default settings 100 % applied!
**** GMOCCAPY GETINIINFO **** 
```

### PR DESCRIPTION
**Pull request by c-morley**
This will allow the user to control the many debug messages sent to the terminal. 
It uses qtvcp's logging library so no reinventing the wheel. 
Other files should probably be converted to use logging too. 

Resolves #2273

### CODE REVIEW

**Comment by hansu** [comment]
Really great Chris, thank you! :+1: :+1: :+1: 
I also had in mind to use qtvcp's logger for gmoccapy, but I just didn't find the time to implement it.
Can you target the PR to this branch: https://github.com/LinuxCNC/linuxcnc/tree/gmoccapy-3-4-2 ?
Then I can add info to the change log and docs and merge to 2.9.

All those removed imports - those are really not necessary?

File: src/emc/usr_intf/gmoccapy/gmoccapy.py [hunk 1]
```diff
@@ -2067,7 +2029,7 @@ def _init_icon_themes(self):
                 selected_index = [icon_theme[1] for icon_theme in valid_icon_themes].index(icon_theme_preference)
                 icon_theme_choice.set_active(selected_index)
             except ValueError:
-                print(f"Warning: preferred icon-theme '{icon_theme_preference}' not found; switching to 'default'.")
+                LOG.warning(f"Warning: preferred icon-theme '{icon_theme_preference}' not found; switching to 'default'.")
```
Comment (hansu) [inline src/emc/usr_intf/gmoccapy/gmoccapy.py:2032]:
```suggestion
                LOG.warning(f"Preferred icon-theme '{icon_theme_preference}' not found; switching to 'default'.")
```

File: src/emc/usr_intf/gmoccapy/gmoccapy.py [hunk 2]
```diff
@@ -5783,12 +5735,58 @@ def _make_hal_pins(self):
...
+    from qtvcp import logger
+    LOG = logger.initBaseLogger('GMoccapy', log_file=None, log_level=logger.WARNING)
+
+    # we set the log level early so the imported modules get the right level
+    if '-d' in sys.argv:
+        # Log level defaults to INFO, so set lower if in debug mode
+        logger.setGlobalLevel(logger.DEBUG)
+        LOG.debug('DEBUGGING logging on')
+    elif '-i' in sys.argv:
+        # Log level defaults to INFO, so set lower if in debug mode
```
Comment (hansu) [inline src/emc/usr_intf/gmoccapy/gmoccapy.py:5758]:
```suggestion
        # Log level defaults to WARNING, so set lower if in debug mode
```
defaults to WARNING if I get it right?

What about a flag for CRITICAL?

**Review by hansu — COMMENTED** [review]
Great job in assigning the log levels! I agree in almost most of the cases, however, I have still some suggestions.

And the correct order of log levels is DEBUG, INFO, VERBOSE, WARNING, ERROR, CRITICAL, is that right?

And they can be only set like this, right?
```
[DISPLAY]
DISPLAY = gmoccapy -q
```

Maybe it would be good to have a LOG_LEVEL in the DISPLAY section as well.

**Comment by c-morley** [comment]
> Great job in assigning the log levels! I agree in almost most of the cases, however, I have still some suggestions.
> 
> And the correct order of log levels is DEBUG, INFO, VERBOSE, WARNING, ERROR, CRITICAL, is that right?
> 
> And they can be only set like this, right?
> 
> ```
> [DISPLAY]
> DISPLAY = gmoccapy -q
> ```
> 
> Maybe it would be good to have a LOG_LEVEL in the DISPLAY section as well.

verbose, debug, info, warning, error, critical
unfortunately at the moment verbose works but does not say 'verbose' it says 'debug' and is the same color as debug.

LOG_LEVEL is more difficult because the log levels must be set before loading some libraries and before instantiating gmoccapy . Not impossible -just have to add code to parse the INI.
 In the current suggestion it is consistent with qtvcp which is nice too.

i don;t know how to change the base branch - could you instruct me?

File: src/emc/usr_intf/gmoccapy/gmoccapy.py (hunk 2 above)
Comment (c-morley) [reply src/emc/usr_intf/gmoccapy/gmoccapy.py:5758]: yes i se the comments are wrong - copied from qtvcp and i changed the default level recently.
I'll need to fix them on qtvcp too. thanks

my opinion:
critical - in my mind if you have critical errors you should pretty much be logging then shutting down.
Having a critical log setting would block errors from showing. can't think of any reason you would want to hide errors.

File: src/emc/usr_intf/gmoccapy/gmoccapy.py [hunk 3]
```diff
@@ -5783,12 +5735,58 @@ def _make_hal_pins(self):
...
+    from qtvcp import logger
+    LOG = logger.initBaseLogger('Gmoccapy', log_file=None, log_level=logger.WARNING)
+
+    # we set the log level early so the imported modules get the right level
+    if '-d' in sys.argv:
+        # Log level defaults to WARNING, so set lower if in debug mode
+        logger.setGlobalLevel(logger.DEBUG)
+        LOG.debug('DEBUGGING logging on')
+    elif '-i' in sys.argv:
+        # Log level defaults to WARNING, so set lower if in debug mode
```
Comment (hansu) [inline src/emc/usr_intf/gmoccapy/gmoccapy.py:5758]:
```suggestion
        # Log level defaults to WARNING, so set lower if in info mode
```

**Comment by hansu** [comment]
> LOG_LEVEL is more difficult because the log levels must be set before loading some libraries and before instantiating gmoccapy . Not impossible -just have to add code to parse the INI. In the current suggestion it is consistent with qtvcp which is nice too.

Yeah but it's fine so. No problem if documented properly.
 
> i don;t know how to change the base branch - could you instruct me?

You'll find the edit button for that at the top of this page:
grafik

**Comment by c-morley** [comment]
Base has been changed, thanks !

**Comment by c-morley** [comment]
I could squash together if you would prefer.

**Comment by hansu** [comment]
> I could squash together if you would prefer.

Yeah would be nice. Have you any thoughts on my other comments?

**Comment by c-morley** [comment]
I thin I have answered all comments - if I have missed one please ask me again or point them out.
If you are referring to some code changes, say changing from error to warning or such - I have no strong opinion - I just set them to what seemed reasonable in a 10 second look.
Feel free to change them to what ever you feel is best.
I'll squish the branch in a couple minutes.

**Comment by c-morley** [comment]
That is the integer representation of the global  log level.
It is in qtvcp/logger.py
I think we should remove it. Opinion?

**Comment by hansu** [comment]
Then it feels a bit redundant, because it says first "INFO logging on" and then this. (Didn't remember if what it prints for default - WARNING).
And using "Base" confused me a bit as well as switching to an integer representation.
So yes can be removed.


Another thing came into my mind while I had a look at the multi-line log messages. It would be nice if the output can be look like this:
```
[Gmoccapy][INFO] A long long long long long long long
                 long long message that goes over several lines.
```
Currently, it's only possible either so
```
[Gmoccapy][INFO] A long long long long long long long
long long message that goes over several lines.
```
or so
```
[Gmoccapy][INFO] A long long long long long long long
[Gmoccapy][INFO] long long message that goes over several lines.
```
First would be no problem if there were no other messages. For the second it's hard so see how many errors/warning there are actually.
(I mean yes you can add the correct number of spaces - the logging domain woul probably not change so that would work, but inconvenient and a bit ugly 🙃)
I thought about something like this, so a function "newline" that adds as many spaces as the last prefix needed:
```
LOG.info(" A long long long long long long long")
LOG.newline("long long message that goes over several lines.")
```

Dou you think that is a reasonable feature?

**Comment by c-morley** [comment]
for LOG.newline()to be useful one would need LOG.info_newline() plus all the other log levels.
I am still having trouble with getting extra log levels to work properly.
you could use:
LOG.info("A long long long long long long long line\n        and the rest of the long line line.")

**Comment by c-morley** [comment]
should be 8 spaces between \n and the 'and'

**Comment by hansu** [comment]
>for LOG.newline()to be useful one would need LOG.info_newline() plus all the other log levels.

I thought about a static class variable that "remembers" the last used log level.

>  you could use:
> LOG.info("A long long long long long long long line\n and the rest of the long line line.")
> should be 8 spaces between \n and the 'and'

Yeah that's what I was meaning in the sentence inside the parantheses.

**Comment by hansu** [comment]
>for LOG.newline()to be useful one would need LOG.info_newline() plus all the other log levels.

I thought about a static class variable that "remembers" the last used log level.

>  you could use:
> LOG.info("A long long long long long long long line\n and the rest of the long line line.")
> should be 8 spaces between \n and the 'and'

Yeah that's what I was meaning in the sentence inside the parantheses.

**Comment by hansu** [comment]
Let's skip this. This seems to be too much work for a bit improvement in readability. Because in the log file there has to be added number of spaces...

**Pull request merged**